test/e2e/k8s_util.go

```diff
@@ -430,12 +430,40 @@ func (k *KubernetesUtils) probeAndDecideConnectivity(fromPod, toPod v1.Pod,
 	fromPodName, toPodName string, port int32, protocol utils.AntreaPolicyProtocol, expectedResult *PodConnectivityMark) (PodConnectivityMark, error) {
 	// Both IPv4 and IPv6 address should be tested.
 	connectivity := Unknown
-	for _, eachIP := range toPod.Status.PodIPs {
-		toIP := eachIP.IP
-		// If it's an IPv6 address, add "[]" around it.
-		if strings.Contains(toIP, ":") {
-			toIP = fmt.Sprintf("[%s]", toIP)
+	var toIPs []string
+	if fromPod.Spec.NodeName == toPod.Spec.NodeName &&
+		!fromPod.Spec.HostNetwork &&
+		toPod.Spec.HostNetwork &&
+		(protocol == utils.ProtocolUDP || protocol == utils.ProtocolSCTP) {
```

Comment:
I wonder if we should use the gateway IP any time the destination Pod is in the hostNetwork. That would simplify things if the only condition is `if toPod.Spec.HostNetwork`. If I understand correctly, this should always work, regardless of source & protocol
Would like to hear @tnqn's opinion here. Maybe he has a better solution for this issue.

Reply:
Always using the gateway IP sounds good to me. 

Another possible solution is to change serve_hostname's implementation and make it use `ListenPacket` instead of `ListenUDP` so it can get the destination IP of a request and use it as the source IP of the response. I think it kind of makes sense for the tool itself, as inconsistent IPs can cause unnecessary problems and shouldn't be a desired behavior.

Reply:
> Another possible solution is to change serve_hostname's implementation and make it use ListenPacket instead of ListenUDP so it can get the destination IP of a request and use it as the source IP of the response. I think it kind of makes sense for the tool itself, as inconsistent IPs can cause unnecessary problems and shouldn't be a desired behavior.

That's the perfect way to resolve this issue. Could we merge this PR first? It might be a long journey to get the new agnhost image included the fix. I can make a PR to the repo later. When K8s releases the new agnhost image included the fix, we could revert this commit.